src/tools/rbd_mirror/image_map/Policy.cc

```diff
@@ -0,0 +1,684 @@
+  assert(ins.second);
+
+  return instance_id;
+}
+
+template <typename I>
+bool Policy<I>::unmap(const std::string &global_image_id) {
+  assert(m_map_lock.is_wlocked());
+
+  std::string instance_id = lookup(global_image_id);
+  if (instance_id == UNMAPPED_INSTANCE_ID) {
+    return false;
+  }
+
+  m_map[instance_id].erase(global_image_id);
+  return true;
+}
+
+template <typename I>
+std::string Policy<I>::remap(const std::string &global_image_id,
+                             const std::string &source_instance_id,
+                             const std::string &target_instance_id) {
+  assert(m_map_lock.is_locked());
+
+  std::string instance_id = lookup(global_image_id);
+  assert(instance_id != UNMAPPED_INSTANCE_ID);
+
+  m_map[target_instance_id].emplace(global_image_id);
+  m_map[source_instance_id].erase(global_image_id);
+
+  return source_instance_id;
+}
+
+template <typename I>
+void Policy<I>::resume_action(const std::set<std::string> &global_image_ids, int r) {
+  dout(20) << dendl;
+
+  bool start_timer = false;
+
+  {
+    RWLock::WLocker map_lock(m_map_lock);
+
+    for (auto const &global_image_id : global_image_ids) {
+      bool schedule = finish_sub_action(global_image_id, r);
+      if (schedule) {
+        schedule_action(global_image_id);
+        start_timer = true;
+      }
+    }
+  }
+
+  if (start_timer) {
+    schedule_update_task();
+  }
+}
+
+template <typename I>
+void Policy<I>::update_image_mapping(const Updates &to_map, const Updates &to_unmap,
+                                     const Updates &to_purge, Context *on_finish) {
+  dout(20) << dendl;
+
+  // default: empty meta policy
+  bufferlist bl;
+  MirrorPolicyData policy_data;
+  ::encode(policy_data, bl);
+  
+  std::map<std::string, cls::rbd::MirrorImageMap> update_mapping;
+  for (auto const &update : to_map) {
+    update_mapping.emplace(update.global_image_id,
+                           cls::rbd::MirrorImageMap(update.instance_id, bl));
+  }
+  for (auto const &update : to_unmap) {
+    update_mapping.emplace(update.global_image_id,
+                           cls::rbd::MirrorImageMap(update.instance_id, bl));
+  }
+
+  std::set<std::string> remove_global_image_ids;
+  for (auto const &update : to_purge) {
+    remove_global_image_ids.emplace(update.global_image_id);
+  }
+
+  start_async_op();
+  UpdateRequest<I> *req = UpdateRequest<I>::create(
+    m_ioctx, std::move(update_mapping), std::move(remove_global_image_ids),
+    create_async_context_callback(m_threads->work_queue, on_finish));
+  req->send();
+}
+
+template <typename I>
+void Policy<I>::send_updates(const std::set<std::string> &global_image_ids) {
+  dout(20) << dendl;
+
+  C_NotifyListener *ctx = new C_NotifyListener(this);
+
+  {
+    RWLock::WLocker map_lock(m_map_lock);
+
+    for (auto const &global_image_id : global_image_ids) {
+      ImageAction::SubActionType sub_action_type = m_image_action.sub_action_start(global_image_id);
```

Comment:
Invoking this method will invoke one of the subaction callbacks -- might be nice if the switch below could be eliminated and instead have those callbacks queue themselves in the set/remove update lists.